pkg/reconciler/broker/broker.go

```diff
@@ -0,0 +1,480 @@
+		r.Recorder.Eventf(b, corev1.EventTypeNormal, topicCreated, "Created PubSub topic %q", topic.ID())
+	}
+
+	b.Status.MarkTopicReady()
+	// TODO(grantr): this isn't actually persisted due to webhook issues.
+	b.Status.TopicID = topic.ID()
+
+	// Check if PullSub exists, and if not, create it.
+	subID := resources.GenerateDecouplingSubscriptionName(b)
+	sub := client.Subscription(subID)
+	subExists, err := sub.Exists(ctx)
+	if err != nil {
+		logger.Error("Failed to verify Pub/Sub subscription exists", zap.Error(err))
+		return err
+	}
+
+	if !subExists {
+		subConfig := gpubsub.SubscriptionConfig{
+			Topic: topic,
+			Labels: map[string]string{
+				"resource":     "brokers",
+				"broker_class": brokerv1beta1.BrokerClass,
+				"namespace":    b.Namespace,
+				"name":         b.Name,
+				//TODO add resource labels, but need to be sanitized: https://cloud.google.com/pubsub/docs/labels#requirements
+			},
+			//TODO(grantr): configure these settings?
+			// AckDeadline
+			// RetentionDuration
+		}
+		// Create a new subscription to the previous topic with the given name.
+		logger.Debug("Creating sub with cfg", zap.String("id", subID), zap.Any("cfg", subConfig))
+		sub, err = client.CreateSubscription(ctx, subID, subConfig)
+		if err != nil {
+			logger.Error("Failed to create subscription", zap.Error(err))
+			b.Status.MarkSubscriptionFailed("CreationFailed", "Subscription creation failed: %w", err)
+			return err
+		}
+		logger.Info("Created PubSub subscription", zap.String("name", sub.ID()))
+		r.Recorder.Eventf(b, corev1.EventTypeNormal, subCreated, "Created PubSub subscription %q", sub.ID())
+	}
+	//TODO update the subscription's config if needed.
+
+	b.Status.MarkSubscriptionReady()
+	// TODO(grantr): this isn't actually persisted due to webhook issues.
+	b.Status.SubscriptionID = sub.ID()
+
+	return nil
+}
+
+func (r *Reconciler) deleteDecouplingTopicAndSubscription(ctx context.Context, b *brokerv1beta1.Broker) error {
+	logger := logging.FromContext(ctx).Desugar()
+	logger.Debug("Deleting decoupling topic")
+
+	// get ProjectID from config or metadata
+	//TODO(grantr) support configuring project in broker config
+	projectID, err := utils.ProjectID("")
+	if err != nil {
+		logger.Error("Failed to find project id", zap.Error(err))
+		return err
+	}
+
+	client, err := r.CreateClientFn(ctx, projectID)
+	if err != nil {
+		logger.Error("Failed to create Pub/Sub client", zap.Error(err))
+		return err
+	}
+	defer client.Close()
+
+	// Delete topic if it exists. Pull subscriptions continue pulling from the
+	// topic until deleted themselves.
+	topicID := resources.GenerateDecouplingTopicName(b)
+	topic := client.Topic(topicID)
+	exists, err := topic.Exists(ctx)
+	if err != nil {
+		logger.Error("Failed to verify Pub/Sub topic exists", zap.Error(err))
+		return err
+	}
+	if exists {
+		if err := topic.Delete(ctx); err != nil {
+			logger.Error("Failed to delete Pub/Sub topic", zap.Error(err))
+			return err
+		}
```

Comment:
It is, but we don't have to check that here. If we get an error here and return it, finalize will be retried. On retry this code won't run because the existence check above will return false.

Reply:
Reconcile failures prevent an update to targets so maybe it's still preferable to handle it? I added similar logic in the CAL source.

Reply:
I created #872 for this. I think the targets update issue will be addressed when it moves to a separate controller for #852.